src/arraytraits.rs

```diff
@@ -216,11 +216,17 @@ pub const ARRAY_FORMAT_VERSION: u8 = 1u8;
 /// Implementation of `ArrayView::from(&S)` where `S` is a slice or slicable.
 ///
 /// Create a one-dimensional read-only array view of the data in `slice`.
+///
+/// **Panics** if the slice length is greater than `isize::MAX`.
 impl<'a, A, Slice: ?Sized> From<&'a Slice> for ArrayView<'a, A, Ix1>
     where Slice: AsRef<[A]>
 {
     fn from(slice: &'a Slice) -> Self {
         let xs = slice.as_ref();
+        assert!(
+            xs.len() <= ::std::isize::MAX as usize,
+            "Slice length must fit in `isize`.",
+        );
```

Comment:
Shouldn't this be conditional and we only check if size_of A is zero?

Reply:
It would be optimal to have this be a "zero cost" conversion in the regular case

Reply:
That's a good question, and the answer isn't entirely clear. The problem is that while `Vec<T>` and `Box<[T]>` promise never to allocate more than `isize::MAX` bytes, it's unclear whether slices created in other ways are allowed to contain more than `isize::MAX` bytes. If slices with non-ZST elements are allowed to have length greater than `isize::MAX` or contain more than `isize::MAX` bytes, we need to perform this check even for non-ZST elements.

The docs for [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) don't say anything about a maximum length or number of bytes, which makes me think that slices with non-ZST elements longer than `isize::MAX` or containing more than `isize::MAX` bytes may be allowed. However, [the current implementation of indexing for slices in the standard library](https://doc.rust-lang.org/src/core/slice/mod.rs.html#2012-2020) *does* assume that the slice contains no more than `isize::MAX` bytes. (It doesn't check whether the offset in bytes will overflow `isize`, but it calls [`slice.as_ptr().add(index)`](https://doc.rust-lang.org/src/core/slice/mod.rs.html#2035) anyway even though the offset in bytes not overflowing `isize` is a safety constraint for [`add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add).)

If you'd prefer, I wouldn't mind assuming, for the time being, that slices can contain no more than `isize::MAX` bytes since the standard library makes this assumption. Then, the check would only be necessary for ZST elements, and I could simplify the `can_index_slice_not_custom` function in the `dimension` module.

I haven't created an issue yet about the "more than `isize::MAX` bytes in slices" question, but a related issue is rust-lang/rust#54857, which discusses the safety of offsetting pointers from empty slices/`Vec`s by zero bytes.

Reply:
Thanks. I'd go with assuming slices can't be longer than isize::MAX. It goes hand in hand with what you are working on in this PR, IMO

Reply:
Okay, I'll add a commit making that assumption.